zipline/utils/cache.py

```diff
@@ -104,11 +100,210 @@ def __init__(self, cache=None):
+            return
+
+        with self.lock:
+            rmtree(self.path)
+
+    def __getitem__(self, key):
+        with self.lock:
+            try:
+                return self.deserialize(self._keypath(key))
+            except UnboundLocalError:
+                # This is how pandas fails if the file doesn't exist! #pandas
+                raise KeyError(key)
+
+    def __setitem__(self, key, value):
+        with self.lock:
+            try:
+                del self[key]
+            except KeyError:
+                pass
+
+            self.serialize(value, self._keypath(key))
+
+    def __delitem__(self, key):
+        with self.lock:
+            try:
+                os.remove(self._keypath(key))
+            except OSError as e:
+                if e.errno == errno.ENOENT:
+                    # raise a keyerror if this directory did not exist
+                    raise KeyError(key)
+                # reraise the actual oserror otherwise
+                raise
+
+    def __iter__(self):
+        return iter(os.listdir(self.path))
+
+    def __len__(self):
+        return len(os.listdir(self.path))
+
+
+class transactional_file(object):
+    """A context manager for managing a temporary file that will be moved
+    to a non-temporary location if no exceptions are raised in the context.
+
+    Parameters
+    ----------
+    final_path : str
+        The location to move the file when committing.
+    *args, **kwargs
+        Forwarded to NamedTemporaryFile.
+
+    Notes
+    -----
+    The file is committed on __exit__ if there are no exceptions.
+    """
+    def __init__(self, final_path, *args, **kwargs):
+        self._tmpfile = NamedTemporaryFile(*args, **kwargs)
+        self._final_path = final_path
+
+    def commit(self):
+        """Sync the temporary file to the final path.
+        """
+        copyfile(self.name, self._final_path)
+
+    def __getattr__(self, attr):
+        return getattr(self._tmpfile, attr)
+
+    def __enter__(self):
+        self._tmpfile.__enter__()
+        return self
+
+    def __exit__(self, *exc_info):
+        if exc_info[0] is None:
+            self.commit()
+        self._tmpfile.__exit__(*exc_info)
+
+
+class transactional_dir(object):
+    """A context manager for managing a temporary directory that will be moved
+    to a non-temporary location if no exceptions are raised in the context.
+
+    Parameters
+    ----------
+    final_path : str
+        The location to move the file when committing.
+    *args, **kwargs
+        Forwarded to tmp_dir.
+
+    Notes
+    -----
+    The file is committed on __exit__ if there are no exceptions.
+    """
+    def __init__(self, final_path, *args, **kwargs):
+        self.name = mkdtemp()
+        self._final_path = final_path
+
+    def commit(self):
+        """Sync the temporary directory to the final path.
+        """
+        copytree(self.name, self._final_path)
```

Comment:
Might be worth noting in the docstring that this isn't truly transactional.  If the commit fails, we don't roll back to the original state.